src/pattern-acceptor.js

```diff
@@ -0,0 +1,532 @@
+/**
+ * Copyright 2018 Shape Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* eslint-disable no-use-before-define */
+
+const syntaxCharacters = '^$\\.*+?()[]{}|'.split('');
+const extendedSyntaxCharacters = '^$\\.*+?()[|'.split('');
+
+const controlEscapeCharacters = 'fnrtv'.split('');
+const controlEscapeCharacterValues = { 'f': '\f'.charCodeAt(0), 'n': '\n'.charCodeAt(0), 'r': '\r'.charCodeAt(0), 't': '\t'.charCodeAt(0), 'v': '\v'.charCodeAt(0) };
+
+const controlCharacters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
+const hexDigits = '0123456789abcdefABCDEF'.split('');
+const decimalDigits = '0123456789'.split('');
+const octalDigits = '01234567'.split('');
+
+
+class PatternAcceptorState {
+  constructor(pattern, unicode) {
+    this.pattern = pattern;
+    this.unicode = unicode;
+    this.index = 0;
+    this.backreferences = [];
+    this.capturingGroups = 0;
+  }
+
+  empty() {
+    return this.index >= this.pattern.length;
+  }
+
+  nextCodePoint() {
+    if (this.empty()) {
+      return false;
+    }
+    return String.fromCodePoint(this.pattern.codePointAt(this.index));
+  }
+
+  skip(n) {
+    for (let i = 0; i < n && this.index < this.pattern.length; i++) {
+      this.index += this.nextCodePoint().length;
+    }
+  }
+
+  eat(str) {
+    if (this.index + str.length > this.pattern.length || this.pattern.slice(this.index, this.index + str.length) !== str) {
+      return false;
+    }
+    this.index += str.length;
+    return true;
+  }
+
+  eatAny(...strs) {
+    for (let str of strs) {
+      if (this.eat(str)) {
+        return str;
+      }
+    }
+    return false;
+  }
+
+  match(str) {
+    return this.index + str.length <= this.pattern.length && this.pattern.slice(this.index, this.index + str.length) === str;
+  }
+
+  matchAny(...strs) {
+    for (let str of strs) {
+      if (this.match(str)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  eatNaturalNumber() {
```

Comment:
Given how this is used, I'd prefer it require non-zero length, and return `null` otherwise (so that you could write `state.eatNaturalNumber() !== null` rather than `state.eatNaturalNumber().length > 0`).